File: js/ClLinearExpression.js
// FILE: EDU.Washington.grad.gjb.cassowary
// package EDU.Washington.grad.gjb.cassowary;

load("mootools-core-1.3.2-server.js");

var ClLinearExpression = new Class({
  /* FIELDS:
     private ClDouble _constant
     private Hashtable _terms
 */
  initialize: function(clv /*ClAbstractVariable*/, value /*double*/, constant /*double*/) {
    if (CL.fGC) print("new ClLinearExpression");
    this._constant = constant || 0;
    this._terms = {};
    if (clv instanceof ClAbstractVariable) this._terms[clv] = value || 1;
    else if (typeof(clv) == 'number') this._constant = clv;
  },

  initializeFromHash: function(constant /*ClDouble*/, terms /*Hashtable*/) {
    if (CL.fGC) print("clone ClLinearExpression");
    this._constant = constant;
    this._terms = Object.clone(terms);
    return this;
  },
  
  multiplyMe: function(x /*double*/) {
    this._constant *= x;
    for (var clv in this._terms) {
      this._terms[clv] *= x;
    }
    return this;
  },

  clone: function() {
    return new ClLinearExpression().initializeFromHash(this._constant, this._terms);
  },

  times: function(x) {
    if (typeof(x) == 'number') {
      return (this.clone()).multiplyMe(x);
    } else {
      if (this.isConstant()) {
        expr = x;
        return expr.times(this._constant);
      } else if (expr.isConstant()) {
        return this.times(expr._constant);
      } else {
        throw new ExCLNonlinearExpression();
      }
    }
  },

  plus: function(expr /*ClLinearExpression*/) {
    if (expr instanceof ClLinearExpression) {
      return this.clone().addExpression(expr, 1.0);
    } else if (expr instanceof ClVariable) {
      return this.clone().addVariable(expr, 1.0);
    }
  },

  minus: function(expr /*ClLinearExpression*/) {
    if (expr instanceof ClLinearExpression) {
      return this.clone().addExpression(expr, -1.0);
    } else if (expr instanceof ClVariable) {
      return this.clone().addVariable(expr, -1.0);
    }
  },


  divide: function(x) {
    if (typeof(x) == 'number') {
      if (CL.approx(x, 0.0)) {
        throw new ExCLNonlinearExpression();
      }
      return times(1.0 / x);
    } else if (x instanceof ClLinearExpression) {
      if (!x.isConstant) {
        throw new ExCLNonlinearExpression();
      }
      return this.times(1.0 / x._constant);
    }
  },

  divFrom: function(expr) {
    if (!this.isConstant() || CL.approx(this._constant, 0.0)) {
        throw new ExCLNonlinearExpression();
    }
    return x.divide(this._constant);
  },

  subtractFrom: function(expr /*ClLinearExpression*/) {
    return expr.minus(this);
  },

  addExpression: function(expr /*ClLinearExpression*/, n /*double*/, subject /*ClAbstractVariable*/, solver /*ClTableau*/) {
    this.incrementConstant(n * expr.constant());
    n = n || 1;
    for (clv in expr.terms()) {
      coeff = expr.terms()[clv];
      this.addVariable(clv, coeff*n, subject, solver);
    }
    return this;
  },

  addVariable: function(v /*ClAbstractVariable*/, c /*double*/, subject, solver) {
    c = c || 1.0;
    if (CL.fTraceOn) CL.fnenterprint("addVariable:" + v + ", " + c);
    coeff = this._terms[v];
    if (coeff) {
      new_coefficient = coeff + c;
      if (CL.approx(new_coefficient, 0.0)) {
        if (solver) {
          solver.noteRemovedVariable(v, subject);
        }
        delete this._terms[v];
      } else {
        this._terms[v] = new_coefficient;
      }
    } else {
      if (!CL.approx(c, 0.0)) {
        this._terms[v] = c;
        if (solver) {
          solver.noteAddedVariable(v, subject);
        }
      }
    }
    return this;
  },

  setVariable: function(v /*ClAbstractVariable*/, c /*double*/) {
    this._terms[v] = c;
    return this;
  },

  anyPivotableVariable: function() {
    if (this.isConstant()) {
      throw new ExCLInternalError("anyPivotableVariable called on a constant");
    } 
    
    for (clv in this._terms) {
      if (clv.isPivotable()) return clv;
    }
    return null;
  },
  
  substituteOut: function(outvar /*ClAbstractVariable*/, expr /*ClLinearExpression*/, subject /*ClAbstractVariable*/, solver /*ClTableau*/) {
    if (this.fTraceOn) CL.fnenterprint("CLE:substituteOut: " + outvar + ", " + expr + ", " + subject + ", ...");
    if (this.fTraceOn) this.traceprint("this = " + this);
    var multiplier = this._terms[outvar];
    delete this._terms[outvar];
    this.incrementConstant(multiplier * expr.constant());
    for (var clv in expr.terms()) {
      var coeff = expr.terms()[clv];
      var old_coeff = this._terms[clv];
      if (old_coeff) {
        var newCoeff = old_coeff + multiplier * coeff;
        if (CL.approx(newCoeff, 0.0)) {
          solver.noteRemovedVariable(clv, subject);
          delete this._terms[clv];
        } else {
          this._terms[clv] = newCoeff;
        }
      } else {
        this._terms[clv] = multiplier * coeff;
        solver.noteAddedVariable(clv, subject);
      }
    }
    if (this.fTraceOn) this.traceprint("Now this is " + this);
  },

  changeSubject: function(old_subject /*ClAbstractVariable*/, new_subject /*ClAbstractVariable*/) {
    this._terms[old_subject] = this.newSubject(new_subject);
  },

  newSubject: function(subject /*ClAbstractVariable*/) {
    if (this.fTraceOn) CL.fnenterprint("newSubject:" + subject);
    
    var reciprocal = 1.0 / this._terms[subject];
    delete this._terms[subject];
    this.multiplyMe(-reciprocal);
    return reciprocal;
  },

  coefficientFor: function(clv /*ClAbstractVariable*/) {
    return this._terms[clv] || 0;
  },

  constant: function() {
    return this._constant;
  },

  set_constant: function(c /*double*/) {
    this._constant = c;
  },

  terms: function() {
    return this._terms;
  },

  incrementConstant: function(c /*double*/) {
    this._constant += c;
  },

  isConstant: function() {
    return Object.keys(this._terms).length == 0;
  },

  toString: function() {
    var bstr = ''; // answer
    var needsplus = false;
    if (!CL.approx(this._constant, 0.0) || this.isConstant()) {
      bstr += this._constant;
      if (this.isConstant()) {
        return bstr;
      } else {
        needsplus = true;
      }
    } 
    for (clv in this._terms) {
      var coeff = this._terms[clv];
      if (needsplus) {
        bstr += " + ";
      }
      bstr += coeff + "*" + clv;
      needsplus = true;
    }
    return bstr;
  },

  Plus: function(e1 /*ClLinearExpression*/, e2 /*ClLinearExpression*/) {
    return e1.plus(e2);
  },
  Minus: function(e1 /*ClLinearExpression*/, e2 /*ClLinearExpression*/) {
    return e1.minus(e2);
  },
  Times: function(e1 /*ClLinearExpression*/, e2 /*ClLinearExpression*/) {
    return e1.times(e2);
  },
  Divide: function(e1 /*ClLinearExpression*/, e2 /*ClLinearExpression*/) {
    return e1.divide(e2);
  },
});

File: java/ClLinearExpression.java
// $Id$
//
// Cassowary Incremental Constraint Solver
// Original Smalltalk Implementation by Alan Borning
// This Java Implementation by Greg J. Badros, <gjb@cs.washington.edu>
// http://www.cs.washington.edu/homes/gjb
// (C) 1998, All Rights Reserved.
//
// (c) 1998 Alan Borning and Greg Badros.  This code is provided for use by
// students for course projects in the course CSE 595 in the Department of
// Computer Science and Engineering, University of Washington, during winter
// quarter 1998.  Any other use requires written permission from the copyright
// holders.
//
// ClLinearExpression
//

import java.util.*;

class ClLinearExpression
{

  public ClLinearExpression(double num)
    {
      my_constant = new Double(num);
      my_terms = new Hashtable();
    }

  public ClLinearExpression(ClAbstractVariable clv, double value, double constant)
    {
      my_constant = new Double(constant);
      my_terms = new Hashtable();
      my_terms.put(clv,new Double(value));
    }

  public ClLinearExpression(Double constant, Hashtable terms)
    {
      my_constant = constant;
      my_terms = terms;
    }


  public ClLinearExpression multiplyMe(double x)
    {
      my_constant = new Double(my_constant.doubleValue() * x);
      
      for (Enumeration e = my_terms.keys() ; e.hasMoreElements(); ) {
        ClVariable clv = (ClVariable) e.nextElement();
	double coeff = ((Double) my_terms.get(clv)).doubleValue();
	my_terms.put( clv, new Double((coeff * x)));
      } 
      return  this;
    }

  public Object clone()
    {
      return new ClLinearExpression(my_constant,my_terms);
    }

  public ClLinearExpression times(double x)
    {
      ClLinearExpression result = (ClLinearExpression) clone();
      return result.multiplyMe(x);
    }

  public ClLinearExpression times(ClLinearExpression expr) throws ExCLNonlinearExpression
    {
      if (isConstant())
	{
	return expr.times(my_constant.doubleValue());
	}
      else if (!expr.isConstant())
	{
	throw new ExCLNonlinearExpression();
	}	
      return times(expr.my_constant.doubleValue());
    }

  public ClLinearExpression plus(ClLinearExpression expr)
    {
      ClLinearExpression result =  this;
      result.addExpression(expr,1.0);
      return result;
    }

  public ClLinearExpression minus(ClLinearExpression expr)
    {
      ClLinearExpression result =  this;
      result.addExpression(expr,-1.0);
      return result;
    }

  public ClLinearExpression divide(double x) throws ExCLNonlinearExpression
    {
      if (ClVariable.clApprox(x,0.0))
	{
	throw new ExCLNonlinearExpression();
	}
      return times(1.0/x);
    }

  public ClLinearExpression divide(ClLinearExpression expr) throws ExCLNonlinearExpression
    {
      if (!expr.isConstant())
	{
	throw new ExCLNonlinearExpression();
	}
      return divide(expr.my_constant.doubleValue());
    }

  public ClLinearExpression divFrom(ClLinearExpression expr) throws ExCLNonlinearExpression
    {
      if (!isConstant() || ClVariable.clApprox(my_constant.doubleValue(),0.0))
	{
	throw new ExCLNonlinearExpression();
	}
      return expr.divide(my_constant.doubleValue());
    }

  public ClLinearExpression subtractFrom(ClLinearExpression expr)
    { return expr.minus( this); }

  // FIXGJB
//   public ClLinearExpression addExpression(ClLinearExpression expr, double n,
// 					  ClAbstractVariable subject, 
// 					  ClTableau solver)
//     {
//       incrementConstant(n * expr.constant());
      
//       for (Enumeration e = expr.terms().keys() ; e.hasMoreElements(); ) {
//         ClVariable clv = (ClVariable) e.nextElement();
// 	double coeff = ((Double) expr.terms().get(clv)).doubleValue();
// 	addVariable(clv,coeff*n, subject, solver);
//       }
//       return this;
//     }

  public ClLinearExpression addExpression(ClLinearExpression expr, double n)
    {
      incrementConstant(n * expr.constant());
      
      for (Enumeration e = expr.terms().keys() ; e.hasMoreElements(); ) {
        ClVariable clv = (ClVariable) e.nextElement();
	double coeff = ((Double) expr.terms().get(clv)).doubleValue();
	addVariable(clv,coeff*n);
      }
      return this;
    }


  public ClLinearExpression addVariable(ClAbstractVariable v, double c)
    { // body largely duplicated below
      //#ifndef CL_NO_TRACE
      //Tracer TRACER(__FUNCTION__);
      //System.err.println("(" + String.valueOf(v) + ", " + String.valueOf(c) + ")");
      //#endif

      Double coeff = (Double) my_terms.get(v);
      if (coeff != null) 
	{
	double new_coefficient = coeff.doubleValue() + c;
	if (ClVariable.clApprox(new_coefficient,0))
	  {
	  my_terms.remove(v);
	  }
	else 
	  {
	  my_terms.put(v,new Double(new_coefficient));
	  }
	}
      else
	{
	if (!ClVariable.clApprox(c,0.0))
	  {
	  my_terms.put(v,new Double(c));
	  }
	}
      return this;
    }

  public ClLinearExpression setVariable(ClAbstractVariable v, double c)
    { 
      //assert(c != 0.0);  
      my_terms.put(v,new Double(c)); return this;
    }

  // FIXGJB  
//   public ClLinearExpression addVariable(ClAbstractVariable v, double c,
// 					ClAbstractVariable subject, ClTableau solver)
//     { // body largely duplicated above
//       //#ifndef CL_NO_TRACE
//       // Tracer TRACER(__FUNCTION__);
//       // System.err.println("(" + String.valueOf(v) + ", " + String.valueOf(c) + ", " + String.valueOf(subject) + ", ...)");
//       // #endif
//       Double coeff = (Double) my_terms.get(v);
//       if (coeff != null) 
// 	{
// 	double new_coefficient = coeff.doubleValue() + c;
// 	if (ClVariable.clApprox(new_coefficient,0))
// 	  {
// 	  solver.noteRemovedVariable(v,subject);
// 	  my_terms.remove(v);
// 	  }
// 	else 
// 	  {
// 	  my_terms.put(v, new Double( new_coefficient));
// 	  }
// 	}
//       else
// 	{
// 	if (!ClVariable.clApprox(c,0.0))
// 	  {
// 	  my_terms.put(v,new Double(c));
// 	  solver.noteAddedVariable(v,subject);
// 	  }
// 	}
//       return this;
//     }
  
  public ClAbstractVariable  anyVariable() throws ExCLInternalError
    {
      if (isConstant())
	{
	throw new ExCLInternalError();
	}
      return (ClAbstractVariable) my_terms.keys().nextElement();
    }

  // FIXGJB  
  //  public void substituteOut(ClAbstractVariable var, ClLinearExpression expr, 
  //			    ClAbstractVariable subject, ClTableau solver)
    {
      //#ifndef CL_NO_TRACE
      //System.err.print("* ClLinearExpression::");
      //Tracer TRACER(__FUNCTION__);
      //System.err.print("(" + String.valueOf(var) + ", " + String.valueOf(expr) + ", " + String.valueOf(subject) + ", ");
      //System.err.println("*this == " + String.valueOf(*this));
      //#endif


      // FIXGJB write this
    }
  
  public void changeSubject(ClAbstractVariable old_subject, ClAbstractVariable new_subject)
    {
      my_terms.put(old_subject,new Double(newSubject(new_subject)));
    }
  
  public double newSubject(ClAbstractVariable subject)
    {
      //#ifndef CL_NO_TRACE
      //Tracer TRACER(__FUNCTION__);
      //System.err.println("(" + String.valueOf(subject) + ")");
      //#endif
      Double coeff = (Double) my_terms.remove(subject);
      // assert(coeff != null);
      double reciprocal = 1.0 / coeff.doubleValue();
      multiplyMe(-reciprocal);
      return reciprocal;
    }

  public double coefficientFor(ClAbstractVariable var)
    { 
      Double coeff = (Double) my_terms.get(var);
      if (coeff != null)
	return coeff.doubleValue();
      else
	return 0.0;
    }

  public double constant()
    { return my_constant.doubleValue(); }

  public void set_constant(double c)
    { my_constant = new Double(c); }

  public Hashtable terms()
    { return my_terms; }

  public void incrementConstant(double c)
    { my_constant = new Double(my_constant.doubleValue() + c); }

  public boolean isConstant()
    { return my_terms.size() == 0; }

  public String toString()
    {
      // FIXGJB: write this
      return "FIXGJB";
    }

  public ClLinearExpression Plus(ClLinearExpression e1, ClLinearExpression e2)
    { return e1.plus(e2); }

  public ClLinearExpression Minus(ClLinearExpression e1, ClLinearExpression e2)
    { return e1.minus(e2); }

  public ClLinearExpression Times(ClLinearExpression e1, ClLinearExpression e2) 
    throws ExCLNonlinearExpression
    { return e1.times(e2); }

  public ClLinearExpression Divide(ClLinearExpression e1, ClLinearExpression e2)
    throws ExCLNonlinearExpression
    { return e1.divide(e2); }

  public boolean FEquals(ClLinearExpression e1, ClLinearExpression e2)
    { return e1 == e2; }

  private Double my_constant;
  private Hashtable my_terms;
  
}
